### PR TITLE
Feature: print tokens per second during training

### LINKED ISSUE #34471: Move Trainer's tokens per second into the inner training loop
### Feature request

Right now `include_tokens_per_second=True` in `Trainer` only reports the tokens per second metric at [the end of training](https://github.com/huggingface/transformers/blob/c1753436dbb8bcbcee183cdd6eba9f08a90d602a/src/transformers/trainer.py#L2610). It would be very useful to have this metric reported continuously inside the training loop so we can monitor it during training. 

### Motivation

The current behavior is counter-intuitive, doesn't align with other convenient trainers (like torchtune), and it's undocumented, so I had to RTFC to figure out why the metric wasn't showing up.

### Your contribution

I probably don't have time to contribute it myself.

### PR DESCRIPTION
**Pull request by tibor-reiss**
Fixes #34471 

If `include_num_input_tokens_seen` was part of `args`, the requested metrics can be printed from `self.state.num_input_tokens_seen`.

@muellerzr and @SunMarc

### CODE REVIEW

**Comment by tibor-reiss** [comment]
I saw #31501 and these two seem related to some extent. Proposal: remove `include_tokens_per_second` and use `num_tokens=self.state.num_input_tokens_seen` instead of `num_tokens=num_train_tokens` when reporting metrics for the full run. What do you think @muellerzr?

**Review by SunMarc — COMMENTED** [review]
We already have the option to set `include_num_input_tokens_seen`. Could you try it and tell us if it fits your use case. One thing that could be nice it to compare the results we get with `include_num_input_tokens_seen` and `include_tokens_per_second` cc @muellerzr

**Comment by tibor-reiss** [comment]
> We already have the option to set `include_num_input_tokens_seen`. Could you try it and tell us if it fits your use case. One thing that could be nice it to compare the results we get with `include_num_input_tokens_seen` and `include_tokens_per_second` cc @muellerzr

Hi @SunMarc, I think we are talking about the same. Someone requested in #34471 the option to be able to see the number of tokens inside the training loop - currently only the total is logged at the end of the training. `num_input_tokens_seen` is already tracked if `include_num_input_tokens_seen`, so I just added one line for the logging. This is the scope of the current PR.

My comment (https://github.com/huggingface/transformers/pull/34507#issuecomment-2447742561) is about simplification, because the current logic poses anyways issues with large datasets. This is not yet covered by this PR, but I could make another one if you agree with my proposal.

**Comment by SunMarc** [comment]
> My comment (https://github.com/huggingface/transformers/pull/34507#issuecomment-2447742561) is about simplification, because the current logic poses anyways issues with large datasets.

Before doing that, could you verify that these two metrics returns actually the same thing. if yes, then we should definitely do as you suggested !

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2472,6 +2471,9 @@ def _inner_training_loop(
                             input_tokens = inputs[main_input_name].numel()
                             input_tokens = torch.tensor(input_tokens, device=self.args.device, dtype=torch.int64)
                             self.state.num_input_tokens_seen += self.accelerator.gather(input_tokens).cpu().item()
+                            self.log(
+                                speed_metrics("train_step", start_time, num_tokens=self.state.num_input_tokens_seen)
+                            )
```
Comment (SunMarc) [inline src/transformers/trainer.py:2476]: Maybe we can put that in `_maybe_log_save_evaluate`  function instead as it is called at the end of each batch

**Comment by tibor-reiss** [comment]
> > My comment ([#34507 (comment)](https://github.com/huggingface/transformers/pull/34507#issuecomment-2447742561)) is about simplification, because the current logic poses anyways issues with large datasets.
> 
> Before doing that, could you verify that these two metrics returns actually the same thing. if yes, then we should definitely do as you suggested !

Results from some basic tests (for speed, I commented out the call to `self.training_step` because it does not influence the result) with the following args:

case 1: `--log_level info     --model_name_or_path=distilbert/distilgpt2    --dataset_name=Salesforce/wikitext     --dataset_config_name=wikitext-2-raw-v1     --block_size=1024     --per_device_train_batch_size=8     --do_train     --output_dir=Llama-3.2-1B-wikitext-2-raw-v1     --overwrite_output_dir     --seed=42     --logging_steps=10     --lr_scheduler_type=cosine     --num_train_epochs=3     --learning_rate=5e-05     --warmup_ratio=0.03     --dataloader_drop_last  --include_tokens_per_second --include_num_input_tokens_seen`

case 2: using different model, i.e. `model_name_or_path=meta-llama/Llama-3.2-1B`

case 3: using `--max_step=50`

In all 3 cases, the values matched:
- case 1: 7102464
- case 2: 7348224
- case 3: 409600

The one test missing is when dataloader does not have a size. Not sure yet how to create that. However, looking at the formula, it is likely that it will also match.

File: src/transformers/trainer.py (hunk 1 above)
Comment (tibor-reiss) [reply src/transformers/trainer.py]: Thanks for the hint, I was not aware of this function. Moved the logic now.

**Comment by LysandreJik** [comment]
cc @SunMarc @muellerzr :)

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -3518,11 +3519,16 @@ def log(self, logs: Dict[str, float]) -> None:
...
+            start_time (`Optional[float]`):
+                The start of training.
         """
         if self.state.epoch is not None:
             logs["epoch"] = self.state.epoch
         if self.args.include_num_input_tokens_seen:
             logs["num_input_tokens_seen"] = self.state.num_input_tokens_seen
+            if start_time is not None:
+                runtime = time.time() - start_time
+                logs["tokens_per_second"] = round(self.state.num_input_tokens_seen / runtime, 3)
```
Comment (SunMarc) [inline src/transformers/trainer.py:3531]: Maybe use `speed_metrics` here and we can also update the description of `include_num_input_tokens_seen` since we also including the speed metrics for tokens per second

**Review by SunMarc — APPROVED** [review]
LGMT ! Thanks for fixing this. A few nits. Can you have a second look @muellerzr ? 
Also @tibor-reiss, it can be a nice follow-up PR to deprecate `include_tokens_per_second`.

File: src/transformers/trainer.py (hunk 2 above)
Comment (tibor-reiss) [reply src/transformers/trainer.py]: Corrected and rebased, thanks!

**Review by ArthurZucker — APPROVED** [review]
🚀 niceeee super long due haha thanks for adding this
